### PR TITLE
Add async dispatch event flag to Chromedriver perform action

### PR DESCRIPTION
**Pull request by LanWei22**
When we dispatch events from chromedriver, we want a flag to specify if we want to dispatch then synchronously or asynchronously. We make the default way of dispatching events as synchronously, when we really need send events asynchronously then we use this flag. Dispatching events asynchronously can send more than one events in one frame.

### CODE REVIEW

**Comment by SummerLW** [comment]
gsnedders@ please take a look, thank you.

**Comment by SummerLW** [comment]
foolip@ Philip, please take a look, thank you.

**Comment by foolip** [comment]
@LukeZielinski can you review this?

File: tools/webdriver/webdriver/client.py [hunk 1]
```diff
@@ -114,17 +114,17 @@ def perform(self):
         """Perform all queued actions."""
         self.session.actions.perform([self.dict])
 
-    def _key_action(self, subtype, value):
-        self._actions.append({"type": subtype, "value": value})
+    def _key_action(self, subtype, value, asyncDispatch=False):
```
Comment (LukeZielinski) [inline tools/webdriver/webdriver/client.py]: Looks like there are several lint errors about the name of this param (eg:
./webdriver/webdriver/client.py:120:49: N803 argument name 'asyncDispatch' should be lowercase)

**Review by LukeZielinski — APPROVED** [review]
lgtm since the change here looks fairly innocuous (% the lint errors about the name of the param).

One clarification: could you provide an example of how this would be used (or link to a relevant bug/doc)?

**Comment by SummerLW** [comment]
The reason we add this attribute to WebDriver action is for some cases we want to dispatch more than one event in one frame, but the currently implementation of WebViewImpl::DispatchMouseEvents is
https://cs.chromium.org/chromium/src/chrome/test/chromedriver/chrome/web_view_impl.cc?type=cs&q=WebViewImpl::DispatchMouseEvents&g=0&l=520
We need a flag to choose which way we want to dispatch events, one is dispatching right away without waiting, the other one is always wait for the response, so that next event be dispatched a frame after the previous one.
For example, this test will need to dispatch asynchronously.
https://cs.chromium.org/chromium/src/third_party/blink/web_tests/external/wpt/pointerevents/extension/pointerevent_coalesced_events_attributes.html

**Comment by LukeZielinski** [comment]
> The reason we add this attribute to WebDriver action is for some cases we want to dispatch more than one event in one frame, but the currently implementation of WebViewImpl::DispatchMouseEvents is
> https://cs.chromium.org/chromium/src/chrome/test/chromedriver/chrome/web_view_impl.cc?type=cs&q=WebViewImpl::DispatchMouseEvents&g=0&l=520
> We need a flag to choose which way we want to dispatch events, one is dispatching right away without waiting, the other one is always wait for the response, so that next event be dispatched a frame after the previous one.
> For example, this test will need to dispatch asynchronously.
> https://cs.chromium.org/chromium/src/third_party/blink/web_tests/external/wpt/pointerevents/extension/pointerevent_coalesced_events_attributes.html

Ok, that explains it. "Multiple events per frame" is the piece I was missing. Thanks!

**Comment by gsnedders** [comment]
This means the WebDriver test client always sends the action objects with a non-standard WebDriver extension, which doesn't seem ideal for the WebDriver tests, especially when the extension appears to be completely undocumented and there's no spec bug open for it or anything…

**Comment by jgraham** [comment]
Yeah, this just seems to be working around a Chrome bug, but in a way that impacts every other implementation. In particular the fact that this affects the wdspec tests means that they are no longer correctly testing the spec.

This change should be reverted; if you need an extra flag for testdriver tests that's possible, but we can't do this for wdspec tests.

**Pull request merged**